src/ConfCrypt/Validation.hs

```diff
@@ -22,71 +22,69 @@ import ConfCrypt.Types
 import ConfCrypt.Encryption (decryptValue, MonadDecrypt)
 
 import Control.Monad.Except (runExcept, catchError)
-import Control.Monad.Writer (MonadWriter, tell)
 import Control.Monad.Reader (MonadReader, ask)
 import Data.Char (isDigit)
-import Data.Foldable (traverse_)
 import Data.Maybe (isNothing)
 import qualified Data.Text as T
 import qualified Data.Map as M
 
 -- | Apply all validation rules, accumulating the errors across rules.
-runAllRules :: (MonadDecrypt m key,
-    Monad m,
-    MonadWriter [T.Text] m,
+runAllRules :: (Monad m,
+    MonadDecrypt m key,
     MonadReader (ConfCryptFile, key) m) =>
-    m ()
+    m [T.Text]
 runAllRules = do
     (ccf, privateKey) <- ask
-    parameterTypesMatchSchema privateKey ccf
-    logMissingSchemas ccf
-    logMissingParameters ccf
+    a <- parameterTypesMatchSchema privateKey ccf
+    b <- logMissingSchemas ccf
+    c <- logMissingParameters ccf
+    pure $ filter (not . T.null) $ a <> b <> c
 
 -- | For each (Schema, Parameter)  pair, confirm that the parameter's value type matches the schema.
-parameterTypesMatchSchema :: (Monad m, MonadWriter [T.Text] m, MonadDecrypt m key) =>
+parameterTypesMatchSchema :: (Monad m, MonadDecrypt m key) =>
     key
     -> ConfCryptFile
-    -> m ()
+    -> m [T.Text]
 parameterTypesMatchSchema key ConfCryptFile {parameters} =
-    traverse_ decryptAndCompare parameters
+    traverse decryptAndCompare parameters
     where
         decryptAndCompare Parameter {paramName, paramValue, paramType} =
             catchError (runRule paramType paramName =<< decryptValue key paramValue)
-                       (const $ tell ["Error: Could not decrypt " <> paramName])
+                       (pure $ pure ("Error: Could not decrypt " <> paramName))
```

Comment:
I already tried but got a mismatch between `m` and `[]`.

Reply:
The second argument of `catchError` is supposed to be a function `e -> m a`, where `a` is instantiated here as `T.Text`.
There is an instance of `Applicative` for functions, where `pure` is implemented as `const`, so this works:

```
f :: Int -> String
f = pure "foo"

f 42 == "foo"
f 5 == "foo"
```

So in this case `pure $ pure "Error ..."` is equivalent to `const $ pure "Error ..."` or `\_ -> pure "Error ..."`.
The first `pure` comes from the `Applicative (a -> b)` instance, and the second pure comes from `m`.

The bottom line is that the first `pure` is technically ok (since it stands for `const`), but confusing, as this comment thread clearly demonstrates.

Reply:
Thanks, it helps. Commited.